Islandora-CLAW/islandora
# Resolves Islandora-CLAW/CLAW#468

In order to install configuration during testing, I had to re-export all of our config and fix some subtle errors that were hard to track down (like adding a schema entry for our config settings).  But on the bright side, we do get our first integration test of sorts in tests/src/Kernel/BroadcasterTest.php.

### To test this PR
* Uninstall the `islandora` module if it is already installed using drush or the web interface
* Pull down the code for this PR
* `$ cd /var/www/html/drupal; composer require stomp-php/stomp-php:4.*` (Once this gets merged, it will become part of the composer install and users will not have to do this manually).
* Install the `islandora` module using drush or the web interface.
* Visit `admin/config/islandora/core` and enter a garbage value for the broker url.  You should get a validation error.
![screenshot from 2017-01-23 14-08-15](https://cloud.githubusercontent.com/assets/20773151/22216930/69335112-e177-11e6-895e-c26d91367df9.png)
* Enter the right value for it: `tcp://localhost:61613`.  You should get a green checkmark.
![screenshot from 2017-01-23 14-08-38](https://cloud.githubusercontent.com/assets/20773151/22216949/775cc05c-e177-11e6-82e2-db54af393a8e.png)
* Visit `admin/config/workflow/rules/reactions/edit/broadcast_create_event` and click 'edit' for the 'Broadcast Message' action.
![screenshot from 2017-01-23 14-12-52](https://cloud.githubusercontent.com/assets/20773151/22216960/8755321e-e177-11e6-95a7-f833a45b35fb.png)
* Enter `activemq:queue:foo` and `activemq:queue:bar`into the 'Recipients' text area and click 'Save'.
![screenshot from 2017-01-23 14-14-11](https://cloud.githubusercontent.com/assets/20773151/22216968/93a39bfa-e177-11e6-9b6f-8d02b464c67f.png)
* Click 'Save' again to save the rule.
* Open up a text editor and save the following xml to `/opt/karaf/deploy/foo.xml`:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<blueprint
        xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xmlns:cm="http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.1.0"
        xsi:schemaLocation="
      http://www.osgi.org/xmlns/blueprint/v1.0.0
      http://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd">

    <camelContext id="IslandoraConnectorBroadcastTest" xmlns="http://camel.apache.org/schema/blueprint">
      <route>
        <from uri="activemq:queue:foo"/>
          <log message="FOO: ${body}"/>
      </route>

      <route>
        <from uri="activemq:queue:bar"/>
          <log message="BAR: ${body}"/>
      </route>

    </camelContext>

</blueprint>
```
* Visit `fedora_resource/add/rdf_source` and create a Fedora resource
![screenshot from 2017-01-23 14-21-41](https://cloud.githubusercontent.com/assets/20773151/22216985/a5af597e-e177-11e6-8e02-d654af984700.png)
* You should see something like this in the karaf logs at `/opt/karaf/data/logs/karaf.log` or `log:tail` from the karaf client:
```
2017-01-23 18:21:21,646 | INFO  | JmsConsumer[foo] | route5                           | 57 - org.apache.camel.camel-core - 2.17.3 | FOO: {"@context":"https:\/\/www.w3.org\/ns\/activitystreams","type":"Create","actor":{"type":"Person","id":"http:\/\/localhost:8000\/user\/1"},"object":"http:\/\/localhost:8000\/fedora_resource\/1"}
2017-01-23 18:21:21,705 | INFO  | ActiveMQ Task-1  | FailoverTransport                | 53 - org.apache.activemq.activemq-osgi - 5.14.0 | Successfully connected to tcp://localhost:61616
2017-01-23 18:21:21,744 | DEBUG | ector-broadcast] | msConfiguration$CamelJmsTemplate | 201 - org.apache.servicemix.bundles.spring-jms - 4.2.5.RELEASE_1 | Executing callback on JMS Session: PooledSession { ActiveMQSession {id=ID:islandora-deux-41238-1485181575873-23:1:1,started=false} java.lang.Object@2d3b36c2 }
2017-01-23 18:21:21,762 | DEBUG | ector-broadcast] | JmsConfiguration                 | 58 - org.apache.camel.camel-jms - 2.17.3 | Sending JMS message to: queue://bar with message: ActiveMQTextMessage {commandId = 0, responseRequired = false, messageId = null, originalDestination = null, originalTransactionId = null, producerId = null, destination = null, transactionId = null, expiration = 0, timestamp = 0, arrival = 0, brokerInTime = 0, brokerOutTime = 0, correlationId = null, replyTo = null, persistent = false, type = null, priority = 4, groupID = null, groupSequence = 0, targetConsumerId = null, compressed = false, userID = null, content = null, marshalledProperties = null, dataStructure = null, redeliveryCounter = 0, size = 0, properties = {breadcrumbId=ID:islandora-deux-46799-1485181588437-4:19:-1:1:1, IslandoraBroadcastRecipients=activemq:queue:foo,activemq:queue:bar, CamelJmsDeliveryMode=1}, readOnlyProperties = false, readOnlyBody = false, droppable = false, jmsXGroupFirstForConsumer = false, text = {"@context":"https:\/\/www.w3.org\/ns\/activi...esource\/1"}}
2017-01-23 18:21:21,798 | DEBUG | ector-broadcast] | MulticastProcessor               | 57 - org.apache.camel.camel-core - 2.17.3 | Done sequential processing 2 exchanges
2017-01-23 18:21:21,822 | DEBUG | JmsConsumer[bar] | faultJmsMessageListenerContainer | 201 - org.apache.servicemix.bundles.spring-jms - 4.2.5.RELEASE_1 | Received message of type [class org.apache.activemq.command.ActiveMQTextMessage] from consumer [PooledMessageConsumer { ActiveMQMessageConsumer { value=ID:islandora-deux-41238-1485181575873-15:1:1:1, started=true } }] of session [PooledSession { ActiveMQSession {id=ID:islandora-deux-41238-1485181575873-15:1:1,started=true} java.lang.Object@658d94a0 }]
2017-01-23 18:21:21,831 | DEBUG | JmsConsumer[bar] | EndpointMessageListener          | 58 - org.apache.camel.camel-jms - 2.17.3 | Endpoint[activemq://queue:bar] consumer received JMS message: ActiveMQTextMessage {commandId = 5, responseRequired = true, messageId = ID:islandora-deux-41238-1485181575873-23:1:1:1:1, originalDestination = null, originalTransactionId = null, producerId = ID:islandora-deux-41238-1485181575873-23:1:1:1, destination = queue://bar, transactionId = null, expiration = 0, timestamp = 1485195681762, arrival = 0, brokerInTime = 1485195681764, brokerOutTime = 1485195681783, correlationId = null, replyTo = null, persistent = true, type = null, priority = 4, groupID = null, groupSequence = 0, targetConsumerId = null, compressed = false, userID = null, content = org.apache.activemq.util.ByteSequence@60c4e6cb, marshalledProperties = org.apache.activemq.util.ByteSequence@116546e1, dataStructure = null, redeliveryCounter = 0, size = 0, properties = {breadcrumbId=ID:islandora-deux-46799-1485181588437-4:19:-1:1:1, CamelJmsDeliveryMode=1, IslandoraBroadcastRecipients=activemq:queue:foo,activemq:queue:bar}, readOnlyProperties = true, readOnlyBody = true, droppable = false, jmsXGroupFirstForConsumer = false, text = {"@context":"https:\/\/www.w3.org\/ns\/activi...esource\/1"}}
2017-01-23 18:21:21,850 | INFO  | JmsConsumer[bar] | route6                           | 57 - org.apache.camel.camel-core - 2.17.3 | BAR: {"@context":"https:\/\/www.w3.org\/ns\/activitystreams","type":"Create","actor":{"type":"Person","id":"http:\/\/localhost:8000\/user\/1"},"object":"http:\/\/localhost:8000\/fedora_resource\/1"}


```
